### PR TITLE
envoy: Add development support for Envoy filter metadata enforcement

### PR DESCRIPTION
**Pull request by jrajahalme**
Introduce a generic mechanism for L7 enforcement on Envoy network
filters that expose meaningful metadata. For an L7 policy to use an
envoy filter, the `l7proto` must be set to the canonical name of the
Envoy network filter, e.g., `envoy.filters.network.mysql_proxy`. L7
policy key/value pairs are mapped to Envoy metadata matchers as
follows:

- `action` key is consumed by the translation and if given with value
  `deny`, will make the rule to drop the connection if there is a
  match.

- Other keys are translated to metadata paths separated by slashes
  (`/`), values are translated to present matches if empty, otherwise
  the value is translated to a list match that matches if the value
  string is one of the values in the metadata.

For example:
```
[{
    "endpointSelector": {"matchLabels":{"mysql-client":""}},
    "egress": [{
        "toPorts": [{
            "ports": [{"port": "3306", "protocol": "TCP"}],
            "rules": {
                "l7proto": "envoy.filters.network.mysql_proxy",
                "l7": [{
                    "action": "deny",
                    "user.mysql": "select"
                }]
            }
        }]
    }]
}]
```
This policy drops the TCP connection established to port 3306 if at
any point the Envoy filter `envoy.filters.network.mysql_proxy` exposes
metadata that has the string `select` in the metadata keyed by
`user.mysql` (`user` table of the `mysql` database).

Policy decisions are access logged translating the filter metadata to
L7 key-value pairs, where the values are JSON formatted. For example:
```
-> Request envoy.filters.network.mysql_proxy from 734 ([container:mysql-client]) to 2341
   ([container:mysql-server]), identity 18656->28273, verdict Forwarded
-> Request envoy.filters.network.mysql_proxy from 734 ([container:mysql-client]) to 2341
   ([container:mysql-server]), identity 18656->28273, verdict Denied user.mysql:["select"]
```
In the first case the filter had not yet exposed any metadata, e.g.,
during MySQL handshake, authentication, or any operation not related
to a table. The second case shows the deny verdict based on the
policy above.

Note that these access log messages are aggregated and are emitted
only when the exposed metadata changes so that multiple operations on
the same table may only emit one access log message.

Note that the Envoy MySQL filter is only used for illustrative
purposes here and it is trivially defeated by allowing the client and
server to negotiate an SSL connection, or using a newer version of
MySQL not fully supported by the Envoy filter (which has been tested
only with MySQL 5.5).

The purpose of this PR is to allow experimentation and development
of more tailored Cilium policy support for the more mature Envoy filters.

### CODE REVIEW

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-only-with-quarantined

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-only-with-quarantined

File: tests/sql-proxy.sh [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+function proxy_init {
+  log "beginning proxy_init"
+  create_cilium_docker_network
+
+  docker run -dt --net=cilium --name mysql-server -l $SERVER_LABEL -e MYSQL_ROOT_PASSWORD=cilium --publish 6603:3306 mysql:$TAG --disable-ssl
+  wait_for_docker_ipv6_addr mysql-server
+
+  log "waiting for all 3 endpoints to get an identity"
+  while [ `cilium endpoint list -o jsonpath='{range [*]}{.status.identity.id}{"\n"}{end}' | grep '^[0-9]' | grep -v '^5$' | wc -l` -ne 3 ] ; do
```
Comment (mrostecki) [inline tests/sql-proxy.sh:42]: Are we assuming that there will be 3 endpoints after running MySQL? On my fresh Vagrant environment I end up having 4 endpoints just after running this test, so I end up having an infinite loop.

Maye we can wait for the mysql endpoint explicitly? ´cilium endpoint list | grep mysql-server`?

Comment (vadorovsky) [inline tests/sql-proxy.sh]: Are we assuming that there will be 3 endpoints after running MySQL? On my fresh Vagrant environment I end up having 4 endpoints just after running this test, so I end up having an infinite loop.

Maye we can wait for the mysql endpoint explicitly?

Comment (vadorovsky) [reply tests/sql-proxy.sh]: This script works for me after this change:

```suggestion
  while ! cilium endpoint list -o jsonpath='{range [*]}{.status.identity.id}{" "}{.status.identity.labels}{"\n"}' | grep '^[0-9].*mysql-server' ; do
```

**Review by vadorovsky — DISMISSED** [review]
Thanks a lot for this PR! Code looks good for me. I have only on suggestion for the bash test.

I will check now whether MongoDB works.

File: tests/sql-proxy.sh (hunk 1 above)
Comment (jrajahalme) [reply tests/sql-proxy.sh]: Fixed, thanks!

**Comment by jrajahalme** [comment]
@mrostecki Small fix and rebase.

**Comment by jrajahalme** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

**Comment by jrajahalme** [comment]
@mrostecki Small fix and rebase.

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-gke

File: pkg/envoy/server.go [hunk 2]
```diff
@@ -285,19 +287,10 @@ func (s *XDSServer) getHttpFilterChainProto(clusterName string, tls bool) *envoy
 	return chain
 }
 
-func (s *XDSServer) getTcpFilterChainProto(clusterName string) *envoy_config_listener.FilterChain {
-	return &envoy_config_listener.FilterChain{
+func (s *XDSServer) getTcpFilterChainProto(clusterName string, useProxylib bool) *envoy_config_listener.FilterChain {
```
Comment (aditighag) [inline pkg/envoy/server.go]: I think it's worthwhile to add a note here indicating the usage of `useProxylib`?

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/envoy/server.go [hunk 3]
```diff
@@ -395,7 +412,43 @@ func (s *XDSServer) AddListener(name string, kind policy.L7ParserType, port uint
...
+					StatPrefix:          "mongo",
+					EmitDynamicMetadata: true,
+				}),
+			}}}, mongoChain.Filters...)
+		mongoChain.FilterChainMatch = &envoy_config_listener.FilterChainMatch{
+			// must have transport match, otherwise TLS inspector will be automatically inserted
+			TransportProtocol:    "raw_buffer",
+			ApplicationProtocols: []string{"envoy.mongo_proxy"},
+		}
+		listenerConf.FilterChains = append(listenerConf.FilterChains, mongoChain)
```
Comment (aditighag) [inline pkg/envoy/server.go]: I wonder if we can extract this common code to a helper function, and pass protocol specific fields to it. That way the code won't be further duplicated as we add more protocols in future.
Something like this - 

```
func prependFilter(FilterChain* , name, typedConfig, ...) {

}
```

File: tests/sql-proxy.sh [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+  log "finished MySQL proxy test"
+}
+
+proxy_init
+
+policy_single_egress
+
+proxy_test
+
+# Leave test setup behind for manual testing
```
Comment (nebril) [inline tests/sql-proxy.sh:125]: Should this stay commented out? I guess we don't really run it anywhere anyway, but what should be the default here?

**Review by nebril — APPROVED** [review]
one question on test script, but since it's not actually run in CI I think I can approve this PR. Thanks!

File: tests/sql-proxy.sh (hunk 4 above)
Comment (jrajahalme) [reply tests/sql-proxy.sh:125]: I'd prefer to leave this commented out, as a reminder that these are left behind. The script is only used for manual testing, so this should not be an issue.

File: pkg/envoy/server.go (hunk 3 above)
Comment (jrajahalme) [reply pkg/envoy/server.go]: Will try this out, thanks!

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-gke

File: pkg/envoy/server.go (hunk 2 above)
Comment (jrajahalme) [reply pkg/envoy/server.go]: Added.

File: pkg/envoy/server.go (hunk 3 above)
Comment (jrajahalme) [reply pkg/envoy/server.go]: Refactored :-)

**Review by vadorovsky — APPROVED** [review]
Looks good. Thanks a lot!

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-gke

**Comment by jrajahalme** [comment]
Unrelated test flake on `netnext`, ready to merge.

**Pull request merged**